File: packages/python/src/softschema/registry.py
"""In-memory collection of registered contracts."""

from __future__ import annotations

from softschema.models import Contract, _check_contract_id


class Contracts:
    """Resolve registered contracts by their public id."""

    def __init__(self) -> None:
        self._contracts: dict[str, Contract] = {}

    def register(self, contract: Contract) -> None:
        existing = self._contracts.get(contract.id)
        if existing is not None and existing != contract:
            msg = f"contract {contract.id!r} is already registered"
            raise ValueError(msg)
        self._contracts[contract.id] = contract

    def resolve(self, contract_id: str) -> Contract | None:
        _check_contract_id(contract_id)
        return self._contracts.get(contract_id)

    @property
    def all(self) -> dict[str, Contract]:
        return dict(self._contracts)

File: packages/typescript/src/registry.ts
/**
 * In-memory collection that resolves contracts by id: the idiomatic mirror of the
 * Python `Contracts`. Registering a different contract under an existing id is an error.
 */
import { type Contract, checkContractId } from "./models.js";

function contractsEqual(a: Contract, b: Contract): boolean {
  return (
    a.id === b.id &&
    a.model === b.model &&
    a.envelopeKey === b.envelopeKey &&
    a.status === b.status &&
    a.profile === b.profile &&
    a.schemaPath === b.schemaPath
  );
}

export class Contracts {
  private readonly contracts = new Map<string, Contract>();

  register(contract: Contract): void {
    checkContractId(contract.id);
    const existing = this.contracts.get(contract.id);
    if (existing !== undefined && !contractsEqual(existing, contract)) {
      throw new Error(`contract ${JSON.stringify(contract.id)} is already registered`);
    }
    this.contracts.set(contract.id, contract);
  }

  resolve(contractId: string): Contract | null {
    checkContractId(contractId);
    return this.contracts.get(contractId) ?? null;
  }

  get all(): Record<string, Contract> {
    return Object.fromEntries(this.contracts);
  }
}
